drivers/storage/portworx/portworx.go

```diff
@@ -189,114 +189,12 @@ func (p *portworx) SetDefaultsOnStorageCluster(toUpdate *corev1.StorageCluster)
-			toUpdate.Status.DesiredImages.LogUploader = release.Components.LogUploader
-		}
-
-		if autoUpdatePxRepo(toUpdate) &&
-			(toUpdate.Status.DesiredImages.PxRepo == "" ||
-				pxVersionChanged ||
-				autoUpdateComponents(toUpdate)) {
-			toUpdate.Status.DesiredImages.PxRepo = release.Components.PxRepo
-		}
-
-		if pxutil.IsCSIEnabled(toUpdate) {
-			if toUpdate.Status.DesiredImages.CSIProvisioner == "" ||
-				pxVersionChanged ||
-				autoUpdateComponents(toUpdate) {
-				toUpdate.Status.DesiredImages.CSIProvisioner = release.Components.CSIProvisioner
-				toUpdate.Status.DesiredImages.CSINodeDriverRegistrar = release.Components.CSINodeDriverRegistrar
-				toUpdate.Status.DesiredImages.CSIDriverRegistrar = release.Components.CSIDriverRegistrar
-				toUpdate.Status.DesiredImages.CSIAttacher = release.Components.CSIAttacher
-				toUpdate.Status.DesiredImages.CSIResizer = release.Components.CSIResizer
-				toUpdate.Status.DesiredImages.CSISnapshotter = release.Components.CSISnapshotter
-				toUpdate.Status.DesiredImages.CSIHealthMonitorController = release.Components.CSIHealthMonitorController
-			}
-			if autoUpdateCSISnapshotController(toUpdate) &&
-				(toUpdate.Status.DesiredImages.CSISnapshotController == "" ||
-					pxVersionChanged ||
-					autoUpdateComponents(toUpdate)) {
-				toUpdate.Status.DesiredImages.CSISnapshotController = release.Components.CSISnapshotController
-			}
-		}
-
-		if toUpdate.Spec.Monitoring != nil && toUpdate.Spec.Monitoring.Prometheus != nil {
-			prometheusVersionChanged := p.hasPrometheusVersionChanged(toUpdate)
-			if toUpdate.Spec.Monitoring.Prometheus.Enabled &&
-				(toUpdate.Status.DesiredImages.PrometheusOperator == "" || pxVersionChanged || prometheusVersionChanged) {
-				toUpdate.Status.DesiredImages.Prometheus = release.Components.Prometheus
-				toUpdate.Status.DesiredImages.PrometheusOperator = release.Components.PrometheusOperator
-				toUpdate.Status.DesiredImages.PrometheusConfigMapReload = release.Components.PrometheusConfigMapReload
-				toUpdate.Status.DesiredImages.PrometheusConfigReloader = release.Components.PrometheusConfigReloader
-			}
-			if toUpdate.Spec.Monitoring.Prometheus.AlertManager != nil &&
-				toUpdate.Spec.Monitoring.Prometheus.AlertManager.Enabled &&
-				(toUpdate.Status.DesiredImages.AlertManager == "" || pxVersionChanged || prometheusVersionChanged) {
-				toUpdate.Status.DesiredImages.AlertManager = release.Components.AlertManager
-			}
-		}
-
-		// Reset the component update strategy if it is 'Once', so that we don't
-		// upgrade components again during next reconcile loop
-		if toUpdate.Spec.AutoUpdateComponents != nil &&
-			*toUpdate.Spec.AutoUpdateComponents == corev1.OnceAutoUpdate {
-			toUpdate.Spec.AutoUpdateComponents = nil
+		// The version from manifest does not match desired version, this may be caused by wrong configMap version.
+		if release.PortworxVersion != "" && toUpdate.Spec.Version != "" && release.PortworxVersion != toUpdate.Spec.Version {
```

Comment:
sorry but in what scenario will this happen? like the version is updated manually from stc spec? I see from line 189 if it's empty, it should be guaranteed to be the same.

Reply:
the scenario is only for px-versions configmap, user updates px-versions configmap, but does not update storageCluster, hence the version mismatches